### DISCUSSION TITLE
App to App Media Sharing

### DISCUSSION THREAD

**Discussion by tytan652** [flatpak/xdg-desktop-portal#1141] [New Portals]
Related to [Goal: Peer-to-peer media sharing](https://github.com/orgs/flatpak/projects/2/views/3?pane=issue&itemId=40363203)

### Case n°1

Applications like [SnekStudio](https://snekstudio.com//) want to be able to share a stream (maybe even various) to other applications (e.g. OBS Studio) alike to what [Spout2](https://github.com/leadedge/Spout2) and [Syphon](https://github.com/Syphon/Syphon-Framework) allows to do on respectively Windows and macOS with almost zero-latency, so mostly "PipeWire + DMA-BUF" for Linux.

We can also point out [Gamescope](https://github.com/ValveSoftware/gamescope) that is a Wayland compositor that want to enable sharing in nested scenarios where a xdp implementation is not a solution (https://github.com/obsproject/obs-studio/pull/9607).

### Case n°2

Others applications want to be able to provide a virtual camera (e.g. OBS Studio) to be used as a camera in other applications.
Some of them already does by using v4l2-loopback which require kernel space API to be used.

As an example, OBS Studio plans to move features to PipeWire (https://github.com/obsproject/obs-studio/discussions/7998), but the actual proposed implementation requires an access to the host PipeWire.

### Case n°3

Also some applications might want to act as virtual microphone (device to send their own audio stream) alike to one of [Easy Effects](https://github.com/wwmm/easyeffects) features but without requiring direct access to the host PipeWire.

## Idea of a solution

Note for the following:
- Shown sequence diagram are not strictly sequence diagram
- I didn't deep dive into how PipeWire nodes should be handled, I might do it at the same time that I work on a proof of concept of case n°1

### For case n°1

For this use-case, we need to enable applications to provide screen cast sources (~~virtual~~) to the ScreenCast portal of the xdp implementation.
So the xdp stack will have to connect to the application point to point D-Bus to make it a screen cast sources provider.

**Note: Name in italic are just temporary name**

![Pseudo-Sequence Diagram for App to App ScreenCast v4 CC-0](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/661731ec-82d6-4494-926b-53b14f59cfd7)

It is also imaginable to simplify the per-application implementation by providing a library that provide most of the basic of the implementation.

#### Thought about audio on those Screencast

Also related to [Goal: Window screencast with audio](https://github.com/orgs/flatpak/projects/2/views/3?pane=issue&itemId=40363336).

Since the application provide its own video stream, the application could also provide a audio one paired with the latter.

This could improve the Gamescope sub-usecase, and maybe also allow to some power users of OBS Studio to "stream" an OBS session to another locally without having to rely to solution like NDI or DeckLink if implemented in OBS. Yes, chained OBS Studio setup like that do exist.

### For case n°2

I didn't give enough thought about it, and like I said earlier I need to dive a little more on how PipeWire works with Portals.

For this one, we need to enable applications to give a video stream to xdp to make it a camera available through the Camera portal.

There is no expectation that the virtual camera can do more than an actual camera, so format-wise it should be like a real camera.

**Reminder: Name in italic are just temporary name**
![Pseudo-Sequence Diagram for App Virtual Camera CC-0](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/6e6ab9dc-53d6-4f83-82a6-cad041de4c7f)

### For case n°3

This cases is alike to the case n°2 but relies on [Goal: Audio portal](https://github.com/orgs/flatpak/projects/2/views/3?pane=issue&itemId=40360215)/#1142.

But unlike case n°2, the naming "virtual microphone" is just used as a figure of speech.

**Comment by Mikenux** [comment]
From a UX point of view.

For case 1, I think there could be two possible experiences:
1) Direct sharing of the streams from the sharer app. The user is presented with an "Open with" type dialog box with apps that support receiving streams from sharers. Permission to share the streams is granted when the user has selected an app and clicked “Open”. In the dialog box, we can use more appropriate phrasing and words than “Open (With)”.
2) Grant access to the streams from the requesting app. The user is presented with the source selector from the app with which to share the streams. If this is secure, sharers can expose their streams without asking the user. If not, apps capable of sharing their streams report that they are able to do so. Then, when the selector is called from the requester, two possibilities, depending on security: (1) send a signal to open compatible apps so that they share their streams, (2) list open compatible apps and send the signal when the app is selected. In short, the ability for an app to provide its content to other apps is done from the selector, since it is the portal.

There may be an option to share or not share audio from a stream.

For case 2, either we display a dialog asking to add the camera (confirming its creation) and it is directly available for other apps in a selector, or it can be the same as in the second experience of case 1 but for a virtual camera.

**Reply by tytan652** [reply]
> Direct sharing of the streams from the sharer app. The user is presented with an "Open with" type dialog box with apps that support receiving streams from sharers.

1. We can't know in advance which app can/will do what since those have access to all portals by design, so no "Open with".
2. The plan for the new portal is to be a "provider" to other portal (ScreenCast, Camera, [Audio](https://github.com/flatpak/xdg-desktop-portal/discussions/1142)) without adding over-complexity to them, only those portals UI/UX will be used including the Access one.

> If this is secure, sharers can expose their streams without asking the user.

This is not how permission and the ScreenCast portal works:
- if the permission is unset, the user must be asked if the app can share
- if it's not a session restoration the user must be asked with a selector.

> There may be an option to share or not share audio from a stream.

Out of scope, those of thought for the quoted goals, this is to talk about in another discussion about adding audio to the ScreenCast portal.

**Reply by Mikenux** [reply]
In the second experience, I talk about implicitly granting access to streams from the relevant portals (called "source selector" because we select sources with them). I mean we want to avoid "Allow %s to provide screencast sources".

**Reply by tytan652** [reply]
> I mean we want to avoid "Allow %s to provide screencast sources".

The application must explicitly have permission separately from and before the ScreenCast dialog, there is no point to avoid it.
If the permission is not given, the portal implementation must not be aware of the provider.

**Reply by Mikenux** [reply]
This is wording that should ideally be avoided, not to not grant permission. Here, it is not common, but it is about displaying in the screencast UI a list of open applications having self-declared (or using any other valid hints) to be able to give their content. Clicking on a listed application means that you are implicitly requesting and giving permission for that application to provide its content. This is, I believe, more or less the same for the OpenURI portal: this dialog is displayed from application A, but you are implicitly giving permission to application B to open a link or file that the application A "has".

Also, I think if the streams are pipewire sources, then they can be presented directly in the screencast UI. I don't think you're granting permission to a portal to access things, because the portal is the one granting access.

**Reply by tytan652** [reply]
I began by saying that the whole permission block in case n°1 is there because I thought allowing explicitly the sharing app to share would a good idea, the permission looks valid only if asked while the sharing is declaring itself to xdp and not later in the xdp impl ScreenCast selection. If various consider it unnecessary with discussions and agreement, it will be removed.

In my opinion, you are just asking for the removal of this block.

> but it is about displaying in the screencast UI a list of open applications having self-declared (or using any other valid hints) to be able to give their content

This portal will not enforce UI/UX guidelines to xdp impls, if they want to show an apps list and then a sources list, or just a sources list. It's up to their design guidelines.
Xdp impls will just be made aware of the sharing apps (if allowed) by xdp, and it will be up to them to show them and their sources with their own UI/UX.

> Clicking on a listed application means that you are implicitly requesting and giving permission for that application to provide its content.

"implicit" is different from "explicit", for me the former has no value as a permission. And like I said before the xdp impl have directly access to the sources list.

> Also, I think if the streams are pipewire sources, then they can be presented directly in the screencast UI.

Nope, in the ScreenCast portal, PipeWire streams can be created after the source happen to be requested by the xdp impl, they don't have to exist beforehand.

**Comment by Mikenux** [comment]
Could you describe case 3 further? Is it, for example, an app that applies voice effects and then shares the resulting audio stream?

**Comment by tytan652** [comment]
**This will used as a sort of changelog if I happen to update my illustrations.**

Changes to Pseudo-Sequence Diagram for App to App ScreenCast between [v1](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/9f1ca752-0bfd-4c53-8649-18e16717fe53) and [v2](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/7876f345-c9ae-4460-aa32-39f762ac6d0d):
- Replaced application interface names
- Removed the "maybe ?"
- Various wording changes
- Changed how how the xdp request application sources

**Reply by tytan652** [reply]
Changes to Pseudo-Sequence Diagram for App to App ScreenCast between [v2](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/7876f345-c9ae-4460-aa32-39f762ac6d0d) and [v3](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/8cbc0cf2-a910-4515-a066-61b00387732b):
- Application ScreenCast Session design is overhauled
  - When submitted as a provider, no session is created but the D-Bus name is watched and the token is stored for later
    - If the D-Bus disappear, the token is destroyed
    - Potential future addition: Checking if the D-Bus name has the needed interface
  - Sources list is given by the application ScreenCast interface (not the session)
    - If the token is wrong, the provider is ignored
  - The Applicatrion ScreenCast Session is created if a application stream happens to be selected

**Reply by tytan652** [reply]
Changes to Pseudo-Sequence Diagram for App to App ScreenCast between [v3](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/8cbc0cf2-a910-4515-a066-61b00387732b) and [v4](https://github.com/flatpak/xdg-desktop-portal/assets/17492366/661731ec-82d6-4494-926b-53b14f59cfd7):
- Move to a point to point D-Bus connection design for the providing app
  - I want to thank Christian Hergert for releasing [Collaborating on Builder](https://blogs.gnome.org/chergert/2024/04/24/collaborating-on-builder/) where I discovered this
  - Just in case, it's Unix Domain Sockets as a transport on the [D-Bus spec](https://dbus.freedesktop.org/doc/dbus-specification.html#transports-unix-domain-sockets)
  - Removes all to *token* complexity
  - Removes all the same App ID or D-Bus name stuff since it's now mostly file descriptors
  
NOTE: I'm working on a POC with point to point to see if it's viable. If it's not, I'll go back to v3 I guess.

**Comment by tytan652** [comment]
PR for case n°1: https://github.com/flatpak/xdg-desktop-portal/pull/1554

**Comment by nlovlyn** [comment]
I'm excited to see this change make get PRed! My use case is under n1 for gamescope and screensharing a game on steamdeck (game-mode) to either Chrome for Jitsi or through the discord client.

**Comment by swick** [comment]
The problem I have with the overall architecture is that it's very complex. If we made pipewire available inside the sandbox and let the portals just control the "permissions" for pipewire, we could clients simply have tag their video sources for consumption of other apps and let pipewire session manager make it available to other clients (if they have the permission).

**Reply by tytan652** [reply]
I will begin with that case n°1 must not make consumer application add new code paths, it has to go through the ScreenCast API at the end.

For producer apps, I'm willing to listen but nobody really discussed about architecture for more than 1 year so unfortunately expect some kind of tenacity from me.

While I agree the architecture is complex, please understand you'll need to be more thorough on your vision/explanation.

And moreover PipeWire is not an **easy** API/ecosystem at all, do not expect me to understand everything right away.

**Comment by BasiqueEvangelist** [comment]
Another usecase for this is letting users stream only one particular Chrome/Firefox/whatever tab, like how it's possible in intra-browser screen capture  
![image](https://github.com/user-attachments/assets/519fe994-0c19-47a7-a7e4-9c9bcee4ebaf)

**Comment by tytan652** [comment]
Idea of improvement for case n°1, make xdp behave like an aggregator of providers act as 1 provider with an interface meant for backends.
This could avoid having backend implements p2p providers handling.

**Reply by tytan652** [reply]
This change where the provider session is managed:
1. End up with ScreenCast `CreateSession` having a scenario where no Request/Response is emitted.
2. Create a portal interface meant for media managed by xdg-desktop-portal accessed by impls another way (here the interface to get all app sources).
3. Throw a `_in_thread_func` function in GTask to send the request to avoid 1. scenario.

**Comment by hoshinolina** [comment]
Just wanted to throw in that I'm also interested, as this change is needed to implement a Spout2-on-Proton/DXVK to PipeWire/OBS bridge reasonably (which would allow the existing Windows Spout2 sender ecosystem to send video to Linux apps when running on Proton). @tytan652 already has some baseline code for the portal/enumeration bits that he shared with me, and I'm building on it to fill in the PipeWire/texture sharing bits to make it work.

**Comment by Conan-Kudo** [comment]
This seems like overkill for this problem? Why not just directly interact with PipeWire? Accessing the PipeWire socket and pushing something through it is well-trodden ground (there's even code in Weston to do this with its `pipewire` backend).

**Reply by tytan652** [reply]
Because it is not allowed, sandboxes will not have access to all nodes.

**Reply by tytan652** [reply]
Also I really want to avoid having application depending on various behavior/API to get a stream when we could keep the list to only ScreenCast and Camera portal.

We could improve things, it's been almost two year that I tried to improve this maybe with #1739 I could move more logic in the WirePlumber module and it's a maybe (I need feedback from xdp devs on that other discussion).

But the project itself has slow development issue, which xdp devs try to solve.

**Reply by tytan652** [reply]
There is also a reason behind the logic of creating the node only when the source had been explicitly requested by an app, imagine a web browser that provide a screencast source for each tab. Is it a good idea to have dozens of unused node created ?

**Reply by jadahl** [reply]
Interacting directly with PipeWire doesn't work with sandboxing, as that require opening up the sandbox. Apps should not be aware of each other, and this portal is explicitly about one app getting the content of another (via PipeWire), which very much requires hand holding from xdg-desktop-portal.

> Also I really want to avoid having application depending on various behavior/API to get a stream when we could keep the list to only ScreenCast and Camera portal.

I agree, having apps being able to get streams from other apps without any more code changes than a "I want app streams too" is a big benefit of this approach.

**Reply by hoshinolina** [reply]
I think we need to be realistic with what is being actually achieved and the pace of development.

The current status is that, even just window sharing with XDG desktop portals is flaky. There is no automatic re-connect when a window is closed and reopened. This works fine on other platforms (Windows), but not with the KDE portal at least. I just had a situation this morning where I was streaming and I had forgotten I reopened an app that was being shared into OBS, and I had no indication that it needed to be re-connected. The image was just frozen and I thought the problem was elsewhere.

I'm all for having this stuff go through portals, but it needs to offer the UX and reliability that you can already get, today, with raw PipeWire nodes and qpwgraph (or similar). This is already how it works with JACK in the audio world, so there is prior art for the "just give the app PW socket access and let the user do whatever" approach.

If we can't achieve the same level of automation and reliability with portals, then we should also have the raw node, and that could be a stepping stone to a more polished portal experience.

One option I thought of is, the portal itself could have an option to just expose a non-permissioned PW sink node, that could then be used for anything the user desires and driven by any app and connection logic. Then you wouldn't have sandbox problems since it's still going through portals, but you wouldn't be blocked on getting all this app2app casting stuff perfectly designed and implemented and reliable. As long as the resume functionality works, the node mode (which probably should have a user-configurable name) would only have to be selected once, and then would work across restarts without any user input, which is what is needed for things to be reliable and automatable.

I think the users driving this app2app functionality are generally power users (the major use cases are video and stream production), and it makes sense to get there first and then build out more polished user experiences for less savvy users, rather than just block everything on an ideal that just might not have enough dev push behind it to be complete any time soon...

**Reply by tytan652** [reply]
> There is no automatic re-connect when a window is closed and reopened. This works fine on other platforms (Windows), but not with the KDE portal at least. I just had a situation this morning where I was streaming and I had forgotten I reopened an app that was being shared into OBS, and I had no indication that it needed to be re-connected. The image was just frozen and I thought the problem was elsewhere.

> If we can't achieve the same level of automation and reliability with portals, then we should also have the raw node, and that could be a stepping stone to a more polished portal experience.

This discussion is mainly about enabling application to share stream to other apps through the **same** flow as monitor and windows (other cases like virtual video/audio devices are low priority).

But it is not about how the flow is unfit to more "professionnal"/"power user" usecase. This part should be discussed elsewhere.

> the portal itself could have an option to just expose a non-permissioned PW sink node

Every node have to be permission controlled to be given to a sandboxed environment.

> I think we need to be realistic with what is being actually achieved and the pace of development.

This is not because the development pace is slow that we should do compromise or do any kind of temporary solution.

**Reply by hoshinolina** [reply]
> Every node have to be permission controlled to be given to a sandboxed environment.

Why? If the user explicitly chooses to expose an unpermissioned node through the portal UI, why shouldn't they be allowed to?

> This is not because the development pace is slow that we should do compromise or do any kind of temporary solution.

I don't think the existence of simpler to implement options suitable for power users is necessarily a "temporary" solution, it's a pragmatic one and might even help get feedback from users on what kinds of flows and use cases are actually in demand, which could help shape the design of a more integrated solution. I would likely still want to use raw PW nodes for some reasons even when the portal stuff is fully fleshed out, because there are use cases (like programmatically swapping sources externally without app involvement) that are unlikely to be supported through portals.

**Reply by tytan652** [reply]
"Path of least resistance" is a problem to be concerned of, if you want to give "power tools" you need to ensure that not every app and user will use those, and ensure that user is informed of the risk (e.g. remind the user every time the app opens).

You need to create a UI/UX that allows the usual flow to exist without making the "power one" the go to.
And this is a hard problem to solve.

**Reply by hoshinolina** [reply]
I don't think reminding the user every time the app opens makes any sense. That would defeat the whole purpose of making it reliable and automatable.

Yes, making a UI/UX to handle a subset of common use cases safely and intuitively is hard, and I'm not saying that shouldn't happen. But clearly we don't all have infinite time and ability to push things forward, so there has to be some pragmatism. "Perfect is the enemy of good."

There is the ability to meet a large set of use cases in a less intuitive way, with less code and development/debate time. And there is the ability to meet an important subset of use cases in a more intuitive way, with more code and development/debate time. Both are important. I don't think we should give up on doing things "the clean way", and I also don't think doing things "the clean way" means we shouldn't also do "the ugly but powerful way".

I know you and others here think punching holes in the sandbox with unrestricted PipeWire access is bad. I personally don't care about the sandbox and would be perfectly happy to give relevant applications full PipeWire access on my systems, but that's just my personal use case. That's why I'm proposing a middle ground, where the application gets an unpermissioned PipeWire node that every *other* app with full PipeWire access can connect to, on request by the user through the portal. So no security is lost for users who do not explicitly choose to invoke this functionality.

**Reply by hoshinolina** [reply]
I have been investigating the permission system. It looks like PipeWire has a special "L" permission that allows a third party to create links between nodes that can't see each other. This means that the client PW node does **not** need to be given any special permissions, rather it's the manager side that needs an extra bit. The only thing the XDG portal side has to allow is a setup with zero streams.

In other words, the only change needed in xdg-desktop-portal is deleting [this](https://github.com/flatpak/xdg-desktop-portal/blob/main/src/screen-cast.c#L985) check. Surely this shouldn't be very controversial. Portal backends could then simply have a UX for such an unconnected node, and then it's up to the user to use a client with full PW access to manage the connection.

This also obviates any security concerns since by definition no security decision is being made by or through the portal. It is up to the user to use PW superpowers to do what they want.

**Reply by tytan652** [reply]
I repeat the scope of this discussion:
> enabling application to share stream to other apps through the **same** flow as monitor and windows

I will not extend the scope of this discussion and like I said:
> it is not about how the flow is unfit to more "professionnal"/"power user" usecase. This part should be discussed elsewhere.

xdg-desktop-portal is an API, so it's behavior is documented.
A screencast session started without stream is a screencast session that is to be closed by the implementation since the implementation failed to start it.

**Reply by hoshinolina** [reply]
OK, fair enough.

From what I've seen of the ScreenCast interface after digging deeper, it is unsuitable for professional use cases as it exists today (in particular the restore functionality, which I just discovered is more flawed than I thought and cannot be relied upon safely).

Since the goal here is to not to touch that interface for sinks at all or make any other changes to it, even simple ones, that would make this use case more reliable/possible, then I don't think this particular discussion is of interest to me any more. I'm sure if this moves forward it will be very nice for casual usage, but I can't in good faith recommend it for production usage when it has such fundamental limitations and discussion of how to improve them is seen as out of scope.
